wandb/sdk/launch/runner/kubernetes_monitor.py

```diff
@@ -13,82 +15,57 @@
+        return f"{self.group}/{self.version}/{self.plural}"
+
+    def __hash__(self) -> int:
+        """Return a hash of the CustomResource."""
+        return hash(str(self))
+
+
+# Maps phases and conditions of custom objects to agent's internal run states.
 CRD_STATE_DICT: Dict[str, State] = {
-    # Starting states.
     "created": "starting",
     "pending": "starting",
-    # Running states.
     "running": "running",
     "completing": "running",
-    # Finished states.
     "succeeded": "finished",
     "completed": "finished",
-    # Failed states.
     "failed": "failed",
     "aborted": "failed",
     "timeout": "failed",
     "terminated": "failed",
-    # Stopping states.
     "terminating": "stopping",
 }
 
-
 _logger = logging.getLogger(__name__)
 
 
-class SafeWatch:
-    """Wrapper for the kubernetes watch class that can recover in more situations."""
-
-    def __init__(self, watcher: "watch.Watch") -> None:
-        """Initialize the SafeWatch."""
-        self._watcher = watcher
-        self._last_seen_resource_version: Optional[str] = None
-        self._stopped = False
-
-    def stream(self, func: Any, *args: Any, **kwargs: Any) -> Any:
-        """Stream the watcher."""
-        while True:
-            try:
-                for event in self._watcher.stream(
-                    func, *args, **kwargs, timeout_seconds=15
-                ):
-                    if self._stopped:
-                        break
-                    # Save the resource version so that we can resume the stream
-                    # if it breaks.
-                    object = event.get("object")
-                    if isinstance(object, dict):
-                        self._last_seen_resource_version = object.get(
-                            "metadata", dict()
-                        ).get("resourceVersion")
-                    else:
-                        self._last_seen_resource_version = (
-                            object.metadata.resource_version
-                        )
-                    kwargs["resource_version"] = self._last_seen_resource_version
-                    yield event
-                # If stream ends after stop just break
-                if self._stopped:
-                    break
-            except urllib3.exceptions.ProtocolError as e:
-                wandb.termwarn(f"Broken event stream: {e}")
-            except ApiException as e:
-                if e.status == 410:
-                    # If resource version is too old we need to start over.
-                    del kwargs["resource_version"]
-                    self._last_seen_resource_version = None
-            except Exception as E:
-                wandb.termerror(f"Unknown exception in event stream: {E}")
-
-    def stop(self) -> None:
-        """Stop the watcher."""
-        self._watcher.stop()
-        self._stopped = True
+def _task_callback(task: asyncio.Task) -> None:
```

Comment:
Should we be displaying these errors in the agent logs rather than logging them to our internal logs?

Also, nit, I think this function name could be more explicit about its purpose